libyang/context.py

```diff
@@ -66,24 +66,25 @@ def __init__(
         if not self.cdata:
             raise self.error("cannot create context")
 
-        search_dirs = []
+        search_dirs = set()
         if "YANGPATH" in os.environ:
-            search_dirs.extend(os.environ["YANGPATH"].strip(": \t\r\n'\"").split(":"))
+            search_dirs.update(os.environ["YANGPATH"].strip(": \t\r\n'\"").split(":"))
         elif "YANG_MODPATH" in os.environ:
-            search_dirs.extend(
+            search_dirs.update(
                 os.environ["YANG_MODPATH"].strip(": \t\r\n'\"").split(":")
             )
         if search_path:
-            search_dirs.extend(search_path.strip(": \t\r\n'\"").split(":"))
+            search_dirs.update(search_path.strip(": \t\r\n'\"").split(":"))
```

Comment:
This means you are not preserving the search path ordering.

I also noted that if yanglib_path is specified, the env vars are ignored. This looks like a mistake.

I saw in the code that `ly_ctx_new` ignores duplicate search paths: https://github.com/CESNET/libyang/blob/v2.0.194/src/context.c#L265-L268

This could be leveraged like this:

```diff
diff --git a/libyang/context.py b/libyang/context.py
index 0b1d146a5383..d1537c5cf036 100755
--- a/libyang/context.py
+++ b/libyang/context.py
@@ -45,8 +45,20 @@ def __init__(
         self.cdata = None
         ctx = ffi.new("struct ly_ctx **")
 
+        search_paths = []
+        if "YANGPATH" in os.environ:
+            search_paths.extend(os.environ["YANGPATH"].split(":"))
+        elif "YANG_MODPATH" in os.environ:
+            search_paths.extend(os.environ["YANG_MODPATH"].split(":"))
+        if search_path:
+            search_paths.extend(search_path.split(":"))
+        if search_paths:
+            search_path = ":".join(search_paths)
+        else:
+            search_path = None
+
         if yanglib_path is None:
-            if lib.ly_ctx_new(ffi.NULL, options, ctx) != lib.LY_SUCCESS:
+            if lib.ly_ctx_new(str2c(search_path), options, ctx) != lib.LY_SUCCESS:
                 raise self.error("cannot create context")
         else:
             if yanglib_fmt == "json":
@@ -66,25 +78,6 @@ def __init__(
         if not self.cdata:
             raise self.error("cannot create context")
 
-        search_dirs = []
-        if "YANGPATH" in os.environ:
-            search_dirs.extend(os.environ["YANGPATH"].strip(": \t\r\n'\"").split(":"))
-        elif "YANG_MODPATH" in os.environ:
-            search_dirs.extend(
-                os.environ["YANG_MODPATH"].strip(": \t\r\n'\"").split(":")
-            )
-        if search_path:
-            search_dirs.extend(search_path.strip(": \t\r\n'\"").split(":"))
-
-        if yanglib_path:
-            return
-
-        for path in search_dirs:
-            if not os.path.isdir(path):
-                continue
-            if lib.ly_ctx_set_searchdir(self.cdata, str2c(path)) != 0:
-                raise self.error("cannot set search dir")
-
     def get_yanglib_data(self, content_id_format=""):
         dnode = ffi.new("struct lyd_node **")
         ret = lib.ly_ctx_get_yanglib_data(self.cdata, dnode, str2c(content_id_format))
```